<?xml version="1.0" encoding="utf-8"?>
<launch>

  <!-- <arg name="server_url" default="$(env CARLA_SERVER)"/> -->
  <!-- Server Connection -->
  <arg name='host' default='localhost' doc='IP to connect to CARLA server (set to &apos;localhost&apos; if running server locally)'/>
  <arg name='port' default='2000' doc='Server port (usually 2000)'/>
  <arg name='passive' default='false' doc='Enable/disable passive mode'/>
  <arg name='synchronous_mode_wait_for_vehicle_control_command' default='false'/>
  <arg name='synchronous_mode' default='true'/>
  <arg name='fixed_delta_seconds' default='0.05'/>
  <arg name='timeout' default='10' doc='Server timeout limit (seconds)'/>

  <!-- Server Environment -->
  <arg name="town" default='Town05_Opt' doc='Map to load on startup (either a predefined CARLA town (e.g. &apos;Town03&apos;), or a OpenDRIVE map file)'/>
  <arg name='all' default='False'/>
  <arg name='decals' default='False'/>
  <arg name='foliage' default='True'/>
  <arg name='parked_vehicles' default='False'/>
  <arg name='particles' default='False'/>
  <arg name='walls' default='True'/>

  <!-- Traffic Generation -->
  <arg name='number_of_vehicles' default='0'/>
  <arg name='spawn_radius' default='100.0'/>
  <arg name='seed' default='0'/>

  <!-- Ego Vehcile -->
  <arg name='role_name' default='ego_vehicle'/>
  <arg name="vehicle_filter" default='vehicle.*'/>
  <arg name="spawn_point_ego_vehicle" default="0,-2.5,1,0,0,0"/>
  <arg name="spawn_sensors_only" default="false"/>
  <arg name="invert_steering" default="false"/>

  <!-- Nodes -->
  <include file="$(find carla_ros_bridge)/launch/carla_ros_bridge.launch">
    <arg name='host' value='$(arg host)'/>
    <arg name='port' value='$(arg port)'/>
    <arg name='town' value='$(arg town)'/>
    <arg name='timeout' value='$(arg timeout)'/>
    <arg name='passive' value='$(arg passive)'/>
    <arg name='synchronous_mode' value='$(arg synchronous_mode)'/>
    <arg name='synchronous_mode_wait_for_vehicle_control_command' value='$(arg synchronous_mode_wait_for_vehicle_control_command)'/>
    <arg name='fixed_delta_seconds' value='$(arg fixed_delta_seconds)'/>
  </include>

  <include file="$(find carla_spawn_objects)/launch/carla_example_ego_vehicle.launch">
    <arg name="objects_definition_file" value='$(find carla_shell_bridge)/launch/objects.json'/>
    <arg name='role_name' value='$(arg role_name)'/>
    <arg name="spawn_point_ego_vehicle" value="$(arg spawn_point_ego_vehicle)"/>
    <arg name="spawn_sensors_only" value="$(arg spawn_sensors_only)"/>
  </include>

  <node pkg="swri_nodelet" type="nodelet" name="carla_interface" args="carla_shell_bridge/CarlaSimulationVehicleInterface standalone">
    <param name="invert_steering" value="$(arg invert_steering)"/>
    <remap from="car_cmd" to="carla/ego_vehicle/vehicle_control_cmd"/>
  </node>

  <node pkg="carla_shell_bridge" type="map_loader.py" name="map_loader" output="screen">
    <param name="host" value="$(arg host)"/>
    <param name="port" value="$(arg port)"/>
    <param name="timeout" value="$(arg timeout)"/>
    <param name="all" value="$(arg all)"/>
    <param name="buildings" value="True"/>
    <param name="decals" value="$(arg decals)"/>
    <param name="foliage" value="$(arg foliage)"/>
    <param name="ground" value="True"/>
    <param name="parked_vehicles" value="$(arg parked_vehicles)"/>
    <param name="particles" value="$(arg particles)"/>
    <param name="street_lights" value="True"/>
    <param name="walls" value="$(arg walls)"/>
  </node>

  <node pkg="carla_shell_bridge" type="traffic_generator.py" name="traffic_generator" output="screen">
    <param name="host" value="$(arg host)"/>
    <param name="port" value="$(arg port)"/>
    <param name="timeout" value="$(arg timeout)"/>
    <param name="number_of_vehicles" value="$(arg number_of_vehicles)"/>
    <param name="spawn_radius" value="$(arg spawn_radius)"/>
    <param name="seed" value="$(arg seed)"/>
  </node>

  <node name="rviz" type="rviz" pkg="rviz" args="-d $(find carla_shell_bridge)/config/default_config.rviz"/>

</launch>
